### ISSUE TITLE
Global JS variables changed by GeckoDriver don't persist

### ISSUE THREAD

**Issue by colons** [mozilla/geckodriver#1067]
## System

* Version: 0.17.0 and 0.19.1
* Platform: Ubuntu 16.04.3 and macOS 10.13.2
* Firefox: 57.0
* Selenium: 3.7.1 with 3.7.0 Python bindings (on Python 2.7 and Python 3.6)

## Testcase

With the following `index.html`:

```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>js desync</title>
  </head>

  <body>
    <h1>hiya</h1>

    <script>
      var variable = "initial";

      function alertVariable() {
        alert(variable);
      }
    </script>
  </body>
</html>
```

And, in the same directory, the following `demo.py`:

```py
#!/usr/bin/env python

import os

from selenium import webdriver


def demo(browser):
    browser.get('file://{}'.format(
        os.path.realpath(os.path.join(os.path.dirname(__file__), 'index.html'))
    ))

    browser.execute_script('variable = "set in script";')
    browser.execute_script('alertVariable();')


if __name__ == '__main__':
    from sys import argv

    browser = getattr(webdriver, argv[-1], None)

    if browser is None:
        raise RuntimeError('could not find a browser called {}'.format(argv[-1]))

    demo(browser())
```

Run `python demo.py Chrome` (or `python demo.py Safari`, or any other supported browser) to observe the intended behaviour; a browser with an alert saying 'set in script'. `python demo.py Firefox`, however, causes an alert saying 'initial'.

This is new behaviour for Firefox, although I'm not sure exactly when it started; the tests this affected in our build hadn't been run for a while when we realised they were failing and worked out that this was the cause.

**Comment by whimboo** [comment]
This is because Marionette is using different sandboxes for each invocation of `execute_script`. As such you do not have access, to such variables from a former sandbox unless you enforce Marionette to re-use such a previously used sandbox.

Interestingly I do not see anything about sandboxes in the WebDriver spec, nor does the webdriver rust module has support for it. @andreastt what is our story with sandboxes nowadays?

**Comment by andreastt** [comment]
> This is because Marionette is using different sandboxes for each invocation of `execute_script`.

This isn’t the case with geckodriver. It uses a sandbox with the following options:

```js
{
  sameZoneAs: window,
  sandboxPrototype: window,
  wantsComponents: false,
  wantsXrays: false,
}
```

This will create a mutable sandbox on which changes will take effect in the DOM. Sandboxes in this context is a Gecko specific concept.

**Comment by andreastt** [comment]
I think if you want to access `variable` above you need to reference it as `window.variable`. @colons, does that work for you?

**Comment by colons** [comment]
Yes; updating the script to do the following instead does indeed make the alert say the right thing.

```py
    browser.execute_script('window.variable = "set in script";')
    browser.execute_script('alertVariable();')
```

The way we worked around this in our tests was to instead store stuff in `data-` attributes of relevant elements, and this definitely seems nicer.

I'm definitely not qualified to comment on whether this means this issue should be closed. All I know is that this is a place where geckodriver behaves differently to other selenium browsers, and it also doesn't mesh at all with my understanding of how javascript scoping is supposed to work.

**Comment by andreastt** [comment]
The injected script is meant to be wrapped on a `function() { … }` (as well as `Promise`, which Marionette doesn’t do yet) and as I understand it the globals of `window` should be accessible to the scope of the function body.

For example:

```js
const foo = "bar";
// foo is now window.foo

function executeScript() {
  return foo;
  // should return window.foo
  // because foo is a global
}
```

**Comment by andreastt** [comment]
I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1420923 to follow up on this.

**Comment by ChristianM18** [comment]
Hi,

I would like to know if anyone could update us regarding this issue?
I found a bug on BugZilla which goal is to "Remove script sandboxing support" (https://bugzilla.mozilla.org/show_bug.cgi?id=1470443). I guess it would fix this issue, wouldn't it?

In my case, I spot the issue when injecting the following script:
```javascript
console.log(this); // => Sandbox{ ... }
console.log(this.window); // => Window
console.log(this.Object === this.window.Object); // false
```

As a workaround, I wrap all the scripts I inject in a JavaScript function, on which I bind the `window` object.
```javascript
console.log(this); // => Sandbox{ ... }
console.log(this.window); // => Window
(function (...args) {
    console.log(this); // => Window
}).call(this.window, ...args);
```

However, this is an ugly workaround, and causes us some other issues. Do you know if there is any way to disable Sandboxing as of today?
